File: checker/course_check/enterprise_capstone.py
"""Deterministic, privacy-preserving checker for the T24 enterprise capstone."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from .evidence import (
    EvidenceError,
    _reject_sensitive_unknown_fields,
    validate_evidence_document,
)


ENTERPRISE_CAPSTONE_LESSON_ID = "t24-enterprise-capstone"
ENTERPRISE_CAPSTONE_VERSION = "1"
ENTERPRISE_CAPSTONE_BASELINE_ID = "telemetry-report-issue-v1"
ENTERPRISE_CAPSTONE_INPUTS = {
    "feature-issue": {"kind": "feature", "title": "报告中增加夜班摘要"},
    "bug-fix": {"kind": "bug", "title": "修复压力单位换算"},
}
ENTERPRISE_CAPSTONE_FAULTS = frozenset(
    {"none", "ambiguous-issue", "test-failure", "review-requested", "mcp-denied"}
)
ENTERPRISE_CAPSTONE_CHECK_IDS = (
    "issue-clarified",
    "context-recorded",
    "memory-governed",
    "skill-applied",
    "mcp-boundary",
    "change-evidence",
    "tests-evidence",
    "review-evidence",
    "delivery-evidence",
    "migration-complete",
    "rubric-complete",
    "offline-deterministic",
)


def _expected_flags(input_id: str, fault: str) -> dict[str, bool]:
    """Return the fixed public status contract for one safe fixture run."""

    ambiguous = fault == "ambiguous-issue"
    return {
        "issue": not ambiguous,
        "context": not ambiguous,
        "memory": True,
        "skill": True,
        "mcp": fault != "mcp-denied",
        "change": not ambiguous,
        "tests": fault != "test-failure" and not ambiguous,
        "review": fault != "review-requested" and not ambiguous,
        "delivery": fault == "none",
        "evidence": True,
        "migration": input_id == "bug-fix" and not ambiguous,
        "rubric": fault == "none" and input_id == "bug-fix",
        "offline": True,
    }


def _checks(flags: dict[str, bool]) -> list[dict[str, str]]:
    fields = (
        ("issue-clarified", "issue"),
        ("context-recorded", "context"),
        ("memory-governed", "memory"),
        ("skill-applied", "skill"),
        ("mcp-boundary", "mcp"),
        ("change-evidence", "change"),
        ("tests-evidence", "tests"),
        ("review-evidence", "review"),
        ("delivery-evidence", "delivery"),
        ("migration-complete", "migration"),
        ("rubric-complete", "rubric"),
        ("offline-deterministic", "offline"),
    )
    return [
        {"id": check_id, "result": "passed" if flags[field] else "failed"}
        for check_id, field in fields
    ]


def load_enterprise_capstone_checks(
    evidence_file: Path,
    *,
    expected_course_version: str,
) -> tuple[list[dict[str, str]], dict[str, Any]]:
    """Validate status-only evidence and rederive its public result.

    The fixture deliberately carries no issue text, source diff, command log,
    account data, path, or model output.  The checker verifies only the
    stable enterprise workflow states and the fixed synthetic fault contract.
    """

    try:
        value = json.loads(evidence_file.read_text(encoding="utf-8"))
    except FileNotFoundError as exc:
        raise EvidenceError(
            f"Enterprise capstone evidence fixture not found: {evidence_file.name}"
        ) from exc
    except json.JSONDecodeError as exc:
        raise EvidenceError(f"Invalid enterprise capstone JSON: {exc.msg}") from exc
    if not isinstance(value, dict):
        raise EvidenceError("Enterprise capstone evidence must be a JSON object")

    # The browser lab may expose a count such as ``12/12 项证据``.  Normalize
    # only this parsed copy to the shared machine-facing evidence summary.
    if value.get("result") == "passed":
        value["summary"] = "所有必需证据均已通过。"
    elif value.get("result") == "partial":
        value["summary"] = "部分证据已通过，仍有证据需要补齐。"
    document = validate_evidence_document(value)
    if document["lesson_id"] != ENTERPRISE_CAPSTONE_LESSON_ID:
        raise EvidenceError(
            "Enterprise capstone fixture lesson_id does not match the requested lesson"
        )
    if document["course_version"] != expected_course_version:
        raise EvidenceError(
            "Enterprise capstone fixture course_version does not match the current course"
        )

    experiment = value.get("experiment")
    if not isinstance(experiment, dict):
        raise EvidenceError("Enterprise capstone evidence requires an experiment object")
    _reject_sensitive_unknown_fields(experiment, "Enterprise capstone experiment")
    if "issue_body" in experiment:
        raise EvidenceError("Enterprise capstone experiment contains a sensitive field: issue_body")
    expected_fields = {
        "version",
        "baseline_id",
        "input",
        "context",
        "memory",
        "skill",
        "mcp",
        "artifacts",
        "migration",
        "rubric",
        "offline",
        "fault",
    }
    if set(experiment) != expected_fields:
        raise EvidenceError("Enterprise capstone experiment fields are invalid")
    if experiment["version"] != ENTERPRISE_CAPSTONE_VERSION:
        raise EvidenceError("Unsupported enterprise capstone experiment version")
    if experiment["baseline_id"] != ENTERPRISE_CAPSTONE_BASELINE_ID:
        raise EvidenceError("Enterprise capstone baseline_id is unsupported")

    input_id = experiment["input"]
    fault = experiment["fault"]
    if input_id not in ENTERPRISE_CAPSTONE_INPUTS:
        raise EvidenceError("Enterprise capstone input variant is unsupported")
    if fault not in ENTERPRISE_CAPSTONE_FAULTS:
        raise EvidenceError("Enterprise capstone fault is unsupported")

    bool_fields = {"context", "memory", "skill", "mcp", "migration", "rubric", "offline"}
    if any(not isinstance(experiment[field], bool) for field in bool_fields):
        raise EvidenceError("Enterprise capstone experiment flags must be booleans")
    artifacts = experiment["artifacts"]
    if not isinstance(artifacts, dict):
        raise EvidenceError("Enterprise capstone artifacts must be an object")
    _reject_sensitive_unknown_fields(artifacts, "Enterprise capstone artifacts")
    artifact_fields = {"change", "tests", "review", "delivery", "evidence"}
    if set(artifacts) != artifact_fields or any(
        not isinstance(artifacts[field], bool) for field in artifact_fields
    ):
        raise EvidenceError(
            "Enterprise capstone artifacts must cover boolean change, tests, review, delivery and evidence"
        )

    expected = _expected_flags(input_id, fault)
    actual = {
        "issue": expected["issue"],
        "context": experiment["context"],
        "memory": experiment["memory"],
        "skill": experiment["skill"],
        "mcp": experiment["mcp"],
        "change": artifacts["change"],
        "tests": artifacts["tests"],
        "review": artifacts["review"],
        "delivery": artifacts["delivery"],
        "evidence": artifacts["evidence"],
        "migration": experiment["migration"],
        "rubric": experiment["rubric"],
        "offline": experiment["offline"],
    }
    if actual != expected:
        raise EvidenceError("Enterprise capstone checks do not match the recorded experiment")

    derived = _checks(actual)
    supplied = document["evidence"]
    if [check["id"] for check in supplied] != list(ENTERPRISE_CAPSTONE_CHECK_IDS):
        raise EvidenceError("Enterprise capstone evidence IDs are invalid")
    if supplied != derived:
        raise EvidenceError("Enterprise capstone checks do not match the recorded experiment")
    expected_result = (
        "passed" if all(check["result"] == "passed" for check in derived) else "partial"
    )
    if document["result"] != expected_result:
        raise EvidenceError("Enterprise capstone result does not match its evidence")
    return derived, {
        "version": ENTERPRISE_CAPSTONE_VERSION,
        "baseline_id": ENTERPRISE_CAPSTONE_BASELINE_ID,
        "input": input_id,
        "context": actual["context"],
        "memory": actual["memory"],
        "skill": actual["skill"],
        "mcp": actual["mcp"],
        "artifacts": {
            field: actual[field]
            for field in ("change", "tests", "review", "delivery", "evidence")
        },
        "migration": actual["migration"],
        "rubric": actual["rubric"],
        "offline": actual["offline"],
        "fault": fault,
    }

File: site/src/lib/enterprise-capstone.mjs
// Deterministic, browser-safe fixture for the T24 enterprise capstone.
// It models public workflow states only; it never reads files, calls a model,
// performs network requests, or writes to an enterprise repository.

export const ENTERPRISE_CAPSTONE_VERSION = "1";
export const ENTERPRISE_CAPSTONE_CHECK_IDS = [
  "issue-clarified",
  "context-recorded",
  "memory-governed",
  "skill-applied",
  "mcp-boundary",
  "change-evidence",
  "tests-evidence",
  "review-evidence",
  "delivery-evidence",
  "migration-complete",
  "rubric-complete",
  "offline-deterministic",
];

const INPUTS = Object.freeze({
  "feature-issue": Object.freeze({
    label: "功能 Issue",
    kind: "feature",
    title: "报告中增加夜班摘要",
  }),
  "bug-fix": Object.freeze({
    label: "缺陷 Issue",
    kind: "bug",
    title: "修复压力单位换算",
  }),
});

const BASELINE = Object.freeze({
  input: "feature-issue",
  context: true,
  memory: true,
  skill: true,
  mcp: true,
  change: true,
  tests: true,
  review: true,
  delivery: true,
  evidence: true,
  migration: false,
  rubric: false,
  offline: true,
  fault: "none",
});

function clone(value) {
  return JSON.parse(JSON.stringify(value));
}
export function inputContract(input = "feature-issue") {
  return clone(INPUTS[input] || INPUTS["feature-issue"]);
}

export function buildEnterpriseRun(overrides = {}) {
  const run = { ...BASELINE, ...overrides };
  if (!Object.hasOwn(INPUTS, run.input)) run.input = "feature-issue";
  const safeFaults = new Set([
    "none",
    "ambiguous-issue",
    "test-failure",
    "review-requested",
    "mcp-denied",
  ]);
  if (!safeFaults.has(run.fault)) run.fault = "none";
  if (run.fault === "ambiguous-issue") {
    run.context = false;
    run.change = false;
    run.tests = false;
    run.review = false;
    run.delivery = false;
    run.migration = false;
  }
  if (run.fault === "test-failure") {
    run.tests = false;
    run.delivery = false;
  }
  if (run.fault === "review-requested") {
    run.review = false;
    run.delivery = false;
  }
  if (run.fault === "mcp-denied") {
    run.mcp = false;
    run.delivery = false;
  }
  if (run.input === "bug-fix" && run.fault !== "ambiguous-issue") {
    run.migration = true;
  }
  if (run.input === "bug-fix" && run.fault === "none") run.rubric = true;

  const checks = [
    ["issue-clarified", run.context],
    ["context-recorded", run.context],
    ["memory-governed", run.memory],
    ["skill-applied", run.skill],
    ["mcp-boundary", run.mcp],
    ["change-evidence", run.change],
    ["tests-evidence", run.tests],
    ["review-evidence", run.review],
    ["delivery-evidence", run.delivery],
    ["migration-complete", run.migration],
    ["rubric-complete", run.rubric],
    ["offline-deterministic", run.offline],
  ].map(([id, passed]) => ({ id, result: passed ? "passed" : "failed" }));

  return {
    version: ENTERPRISE_CAPSTONE_VERSION,
    id: "enterprise-run",
    input: run.input,
    variant: inputContract(run.input),
    fault: run.fault,
    context: Boolean(run.context),
    memory: Boolean(run.memory),
    skill: Boolean(run.skill),
    mcp: Boolean(run.mcp),
    artifacts: {
      change: Boolean(run.change),
      tests: Boolean(run.tests),
      review: Boolean(run.review),
      delivery: Boolean(run.delivery),
      evidence: Boolean(run.evidence),
    },
    migration: Boolean(run.migration),
    rubric: Boolean(run.rubric),
    offline: Boolean(run.offline),
    checks,
  };
}

export function summarizeEnterpriseRun(run) {
  const checks = Array.isArray(run?.checks) ? run.checks : [];
  const passed = checks.filter((check) => check.result === "passed").length;
  return {
    passed,
    total: ENTERPRISE_CAPSTONE_CHECK_IDS.length,
    result: passed === ENTERPRISE_CAPSTONE_CHECK_IDS.length ? "passed" : "partial",
    label: `${passed}/${ENTERPRISE_CAPSTONE_CHECK_IDS.length} 项证据`,
  };
}

export function buildEnterpriseEvidence(run = buildEnterpriseRun()) {
  const normalized = run?.version === ENTERPRISE_CAPSTONE_VERSION
    ? run
    : buildEnterpriseRun(run);
  return {
    contract: "agent-engineering-course/evidence",
    contract_version: "1",
    course_version: "3.0.0",
    lesson_id: "t24-enterprise-capstone",
    result: summarizeEnterpriseRun(normalized).result,
    anonymous: true,
    checked_on: "2026-08-13",
    summary: summarizeEnterpriseRun(normalized).result === "passed"
      ? "所有必需证据均已通过。"
      : "部分证据已通过，仍有证据需要补齐。",
    evidence: normalized.checks,
    experiment: {
      version: ENTERPRISE_CAPSTONE_VERSION,
      baseline_id: "telemetry-report-issue-v1",
      input: normalized.input,
      context: normalized.context,
      memory: normalized.memory,
      skill: normalized.skill,
      mcp: normalized.mcp,
      artifacts: normalized.artifacts,
      migration: normalized.migration,
      rubric: normalized.rubric,
      offline: normalized.offline,
      fault: normalized.fault,
    },
  };
}
